Automate-774 Service groups search/filter bar

tarablack01 opened the pull request:
### :nut_and_bolt: Description: What code changed, and why?
Given the limitation of getting only a services page for MVP, users could not complete basic tasks such as checking the status of one deployment (an application deployed in an environment), or one service across all the environment and applications.

API TODO: service_groups_health_counts api needs to update counts on search bar filters, do we need to apply filter params as well?:
![Screen Shot 2019-08-18 at 2 49 51 PM](https://user-images.githubusercontent.com/4108100/63229722-18319280-c1c9-11e9-8a89-d22966dc67c0.png)
![Screen Shot 2019-08-18 at 2 49 58 PM](https://user-images.githubusercontent.com/4108100/63229723-1b2c8300-c1c9-11e9-9138-9a12da0ab2c4.png)

To fulfill the needs, we would like to provide a search bar with filters that allow users to filter the services table to only the things they care about for a specific task.

### :chains: Related Resources
https://github.com/chef/automate/issues/774
https://chef.invisionapp.com/share/D3T4FUT26B7#/screens/375625354

### :+1: Definition of Done
Reuse the implementation of the search bar/filters from the Client Runs page. This is an explicit requirement because we want a unified search/filter experience across all of Automate based on a single codebase.

We expect to have the following filters in order in the dropdown list:

- Origin, e.g., "core"
- Service Name, e.g., "nginx"
- Version, e.g., "1.2.3"
- Channel
- Application (note that Deployment == Environment + Application)
- Environment
- Site
- Health Check Message
- Build Timestamp, e.g., "20190318112388"
- Group Name (service group name suffix after the '.')

Saving & Sharing filters: append the applied filters to the URL like what Chef Infra search bar does, add the "share" functionality which allows copying the URL with the applied filters saved. In reverse, when updating the filters and values in the URL and refreshing the page, the applied filters update and match with what's in the URL.

Development tooling such as the applications_populate_database function in the studio should be updated as necessary to support the full demonstration of this feature.

### :athletic_shoe: How to Build and Test the Change
build components/automate-ui && start_all_services && applications_populate_database
Go to https://a2-dev.test/applications and inspect

### :white_check_mark: Checklist

- [x ] Tests added/updated?
- [x ] Docs added/updated?

### :camera: Screenshots, if applicable
![search-bar-service-groups](https://user-images.githubusercontent.com/4108100/63394345-e9f8b200-c384-11e9-93f2-f3f8fce0227a.gif)
![search-bar-event-feed](https://user-images.githubusercontent.com/4108100/63394352-ef55fc80-c384-11e9-8da8-d26e33c665dc.gif)
![search-bar-client-runs](https://user-images.githubusercontent.com/4108100/63394353-f250ed00-c384-11e9-899e-0f60009d1f0a.gif)

afiune commented on components/automate-ui/src/app/entities/service-groups/service-groups.requests.ts:
```diff
@@ -79,4 +95,33 @@ export class ServiceGroupsRequests {
+  public getSuggestions(type: string, text: string, filters: NodeFilter): Observable<any[]> {
+    if (text && text.length > 0) {
+      const params = this.formatFilters(filters).set('type', type).set('text', text);
+      const url = `${CONFIG_MGMT_URL}/suggestions`;
```
I might totally be missing something here but, why do we get the suggestions from the `config-mgmt-service`?

tarablack01 replied:
@afiune thanks for pointing that out! I was able to confirm the correct api: applications_url: '/apis/beta/applications', `${APPLICATIONS_URL}/services-distinct-values`

afiune commented on components/automate-ui/src/app/entities/service-groups/service-groups.actions.ts:
```diff
@@ -21,7 +21,11 @@ export enum ServiceGroupsActionTypes {
+
+  GET_NODE_SUGGESTIONS = 'GET_NODE_SUGGESTIONS::SERVICES::GET',
+  GET_NODE_SUGGESTIONS_SUCCESS = 'GET_NODE_SUGGESTIONS::SERVICES::SUCCESS',
+  GET_NODE_SUGGESTIONS_FAILURE = 'GET_NODE_SUGGESTIONS:SERVICES::FAILURE'
```
Should we rename these actions? or why do we have the word "node" in the applications tab? 🤔

tarablack01 replied:
@afiune Updated name `GET_NODE_SUGGESTIONS` to be specific to service groups: `GET_SERVICE_GROUP_SUGGESTIONS`

afiune commented on components/automate-ui/src/app/entities/service-groups/service-groups.actions.ts:
```diff
@@ -87,6 +91,23 @@ export class GetServicesBySGFailure implements Action {
   constructor(public payload: HttpErrorResponse) { }
 }
 
+export class GetNodeSuggestions implements Action {
```
all the classes are still named `GetNode*`

tarablack01 commented:
> I think I am totally out of sync from the plans for the search bar for the applications service, I don't understand why are we adding node actions and client-run requests, also I don't know why are we getting the suggestions from the `config-mgmt-service`. **If this is the plan then ignore me.** 😅 Otherwise, I think the main objective of this Issue/Task was to re-use the search bar components but implement new actions, requests, etc (redux stuff) for the applications service.

@afiune You're correct. I didn't have all the details for the api yet. Now I do. It's all implemented and ready to test 👍🏻

afiune commented on components/automate-ui/src/app/entities/service-groups/service-groups.actions.ts:
```diff
@@ -21,7 +21,11 @@ export enum ServiceGroupsActionTypes {
-  GET_SERVICES_BY_SERVICE_GROUP_FAILURE = 'SERVICE_GROUPS::SERVICES::GET::FAILURE'
+  GET_SERVICES_BY_SERVICE_GROUP_FAILURE = 'SERVICE_GROUPS::SERVICES::GET::FAILURE',
+
+  GET_SERVICE_GROUPS_SUGGESTIONS = 'GET_SERVICE_GROUPS_SUGGESTIONS::SERVICES::GET',
```
To follow the name standard above I would suggest calling it like this:
```suggestion
  GET_SERVICE_GROUPS_SUGGESTIONS = 'SERVICE_GROUPS::SUGGESTIONS::GET',
```
Same for the two below 👇

tarablack01 replied:
@afiune updated 👍🏻

afiune commented on components/automate-ui/src/app/entities/service-groups/service-groups.requests.ts:
```diff
@@ -1,15 +1,23 @@
 import { environment } from '../../../environments/environment';
 const APPLICATIONS_URL = environment.applications_url;
 
+interface RespSuggestion {
```
can we move this definition to the `service-groups.model.ts` file?

tarablack01 replied:
@afiune updated 👍🏻

bcmdarroch commented on components/automate-ui/src/app/pages/service-groups/service-groups.component.ts:
```diff
@@ -167,13 +261,27 @@ export class ServiceGroupsComponent implements OnInit, OnDestroy {
       .subscribe(currentPage => this.currentPage = currentPage);
+
+    this.nodeSuggestions$ = this.store.select(createSelector(serviceGroupState,
+      (state) => state.nodeSuggestions)).pipe(map((nodeSuggestions: any[]) =>
```
could we use a stronger type than `any` here?

bcmdarroch commented on components/automate-ui/src/app/pages/service-groups/service-groups.component.ts:
```diff
@@ -200,16 +308,82 @@ export class ServiceGroupsComponent implements OnInit, OnDestroy {
+    this.shareDropdownVisible = false;
+  }
+
+  onSuggestValues(event) {
```
what types of events are we expecting in these functions?

bcmdarroch commented on components/automate-ui/src/app/pages/service-groups/service-groups.component.ts:
```diff
@@ -75,7 +75,7 @@ export class ServiceGroupsComponent implements OnInit, OnDestroy {
 
   // autocomplete suggestions
-  nodeSuggestions$: Observable<any[]>;
+  serviceGroupsSuggestions$: Observable<any[]>;
```
should these all observable string arrays?

msorens replied:
Yes, I would second the suggestion to get rid of `any` for a specific type. (If you search through your code changes, wherever you reference suggestions it uses `any`, so when changing here, please check for the rest, too.) Thx!

bcmdarroch commented:
Oh and also wanted to mention, a rebase should hopefully take care of that failing Cypress test 🤞 Ping auth team if it doesn't though

msorens commented on components/automate-ui/src/app/entities/service-groups/service-groups.requests.ts:
```diff
@@ -1,11 +1,16 @@
+  ServicesPayload,
+  ServicesFilters
 } from './service-groups.model';
 import { environment } from '../../../environments/environment';
```
Another instance of that path convention I mentioned above. 😁 
```suggestion
import { environment } from 'environments/environment';
```

msorens commented on components/automate-ui/src/app/entities/service-groups/service-groups.requests.ts:
```diff
@@ -79,4 +90,34 @@ export class ServiceGroupsRequests {
+
+  public getSuggestions(
+    field_name: string,
+    query_fragment: string,
```
For consistency should use camel case instead of snake case, so `fieldName` and `queryFragment`. (I think it is OK that they are not identical to the URL params.)

susanev commented:
@tarablack01 havent had time to run it locally yet, but from ur screenshot, the space between the subheading and the search bar doesn't look like it matches the rest of the pages. also, @apriofrost  should there be that gap between the search bar and the save button?

msorens commented on components/automate-ui/src/app/entities/service-groups/service-groups.requests.ts:
```diff
@@ -79,4 +90,34 @@ export class ServiceGroupsRequests {
+  }
+
+  private flattenSearchBar(filters: object[], searchParam: HttpParams): HttpParams {
+    return reduce((params: HttpParams, filter: { type: string, text: string }) => {
```
This filter type is used in about 8 places in the code base now. Would like to see you promote that to an exported interface so these couple lines could be a bit more clearly written:
```
  private flattenSearchBar(filters: FilterType[], searchParam: HttpParams): HttpParams {
    return reduce((params: HttpParams, filter: FilterType) => {
```

msorens commented on components/automate-ui/src/app/page-components/search-bar/search-bar.component.ts:
```diff
@@ -310,6 +310,8 @@ export class SearchBarComponent implements OnChanges {
       this.suggestionsVisible = true;
       this.isLoadingSuggestions = false;
+    } else {
+      this.suggestValues.emit({ detail: { text: '', type: type.type } });
```
With that `FilterType` I mentioned above, you could then use that to enforce strong typing on this line. Right now, if you make a typo--e.g. change `text` to `txet`--the compiler will not complain. But with strong typing it can.

msorens commented on components/automate-ui/src/app/pages/service-groups/service-groups.component.ts:
```diff
@@ -1,14 +1,19 @@
+  RollupServiceStatus,
+  SearchBarCategoryItem,
+  SortDirection
+} from '../../types/types';
```
OK, I won't say anything about `../..` anymore. 😁

msorens commented on components/automate-ui/src/app/pages/service-groups/service-groups.component.ts:
```diff
@@ -63,6 +68,85 @@ export class ServiceGroupsComponent implements OnInit, OnDestroy {
+  // Should the URL share dropdown be displayed
+  shareDropdownVisible = false;
+
+  // Should the search bar filter bar be displayed
```
Is that correct--two uses of the word "bar" there...?

msorens commented on components/automate-ui/src/app/pages/service-groups/service-groups.component.ts:
```diff
@@ -63,6 +68,85 @@ export class ServiceGroupsComponent implements OnInit, OnDestroy {
+  // autocomplete suggestions
+  serviceGroupsSuggestions$: Observable<any[]>;
+
+  // The catagories allowed for searching
```
```suggestion
  // The categories allowed for searching
```

msorens commented on components/automate-ui/src/app/pages/service-groups/service-groups.component.ts:
```diff
@@ -200,16 +304,82 @@ export class ServiceGroupsComponent implements OnInit, OnDestroy {
     this.store.dispatch(new UpdateServiceGroupFilters({filters: serviceGroupFilters}));
   }
 
+  get shareUrl() {
```
```suggestion
  get shareUrl(): string {
```

And while you are in this file, if you don't mind, most of the functions in this file are in need of a `: void` showing no return type. Just a few functions here have it.

tarablack01 commented:
@bcmdarroch @msorens Dang you both went to town on this PR, thanks!! Good job 👍🏻 I'll implemented the requested changes.

tarablack01 commented:
> Interaction-wise, it looks good! A few feedback:
> 
> 1. Spacing change:
>    ![image](https://user-images.githubusercontent.com/9401459/63618708-944d2100-c5a1-11e9-9d69-47119a95637c.png)
> 2. After a filter is selected from the filter list, could we add a loading state before the suggestion list is shown? In case of loading the suggestion list takes some time.
>    ![image](https://user-images.githubusercontent.com/9401459/63618881-ff96f300-c5a1-11e9-872d-3c2d79e58005.png)
> 3. This is a nice to have, doesn't have to be included in this ticket. Wondering if we can make the list align with the focus area of the input field:
>    ![image](https://user-images.githubusercontent.com/9401459/63618953-3ec54400-c5a2-11e9-86e3-04013f85d53c.png)

@apriofrost 1. has been updated 2. already exists 3. We won't do now but keep in mind the left position will always change based on the width of the first selected category; makes it a bit more tricky.

tarablack01 commented:
@apriofrost @bcmdarroch @susanev @msorens 
Implemented requested PR changes, some of the requested changes for this PR were changes that I was making in my next story for refactoring the service groups files and adding a service facade. Since I had already started on that story during this PR I incorporated some of those changes in this PR because the same files overlapped and. The facade and service-groups.component will be refactored in that PR.

tarablack01 commented:
@apriofrost 
I added 32px margin between subheading and search input field.
![Screen Shot 2019-09-09 at 10 57 42 AM](https://user-images.githubusercontent.com/4108100/64546691-c2946580-d2f0-11e9-88bb-2dc15f0ffaae.png)

apriofrost commented:
Looking great! @tarablack01
